Tensor Parallelism conceptual guide (#886)
huggingface/text-generation-inference: docs/source/conceptual/tensor_parallelism.md

# Tensor Parallelism

> Tensor Paralellism (also called horizontal model paralellism) is a technique used to fit a large model in multiple GPUs. Model parallelism enables large model training and inference by putting different layers in different GPUs (also called ranks). Intermediate outputs between ranks are sent and received from one rank to another in a synchronous or asynchronous manner. When multiplying input with weights for inference, multiplying input with weights directly is equivalent to dividing weight matrix column-wise, multiplying each column with input separately, and then concatenating the separate outputs like below 👇 

Reviewer: * Typo in a couple of places: Parallelism -> Parallelism
* We discuss tensor parallelism (horizontal), but the explanation corresponds to pipeline parallelism (vertical). As per my understanding, tensor parallelism sends parts of a tensor to different GPUs (not layers). I would just explain tensor parallelism in this guide to focus on tensor parallelism (https://colossalai.org/docs/concepts/paradigms_of_parallelism/#tensor-parallel helped me here)

Author: I explained model parallelism first and since tensor parallelism is sub-topic of this I thought I'd mention that.

Author: I removed MP.

![Image courtesy of Anton Lozkhov](https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/tgi/TP.png)

In TGI, tensor parallelism is implemented under the hood by sharding weights and placing them in different ranks. The matrix multiplications then take place in different ranks and are then gathered into single tensor. 

<Tip warning={true}>

> Tensor Parallelism only works for model with custom kernels.

Reviewer 2: ```suggestion
Tensor Parallelism only works for model officially supported, it will not work when falling back on `transformers`.
```

> </Tip>

Reviewer: If we aim to be minimal, I would add a section with links to additional resources. E.g. https://huggingface.co/docs/transformers/main/en/perf_train_gpu_many#tensor-parallelism would be great here

# Tensor Parallelism

> Tensor parallelism is a technique used to fit a large model in multiple GPUs. For example, when multiplying the input tensors with the first weight tensor, multiplying both tensors is equivalent to splitting the weight tensor column-wise, multiplying each column with input separately, and then concatenating the separate outputs. These outputs are then sent between GPUs and then concatenated together to get the final result, like below 👇 

Reviewer 3: ```suggestion
Tensor parallelism is a technique used to fit a large model in multiple GPUs. For example, when multiplying the input tensors with the first weight tensor, the matrix multiplication is equivalent to splitting the weight tensor column-wise, multiplying each column with the input separately, and then concatenating the separate outputs. These outputs are then transferred from the GPUs and concatenated together to get the final result, like below 👇 
```

![Image courtesy of Anton Lozkhov](https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/tgi/TP.png)

<Tip warning={true}>

> Tensor Parallelism only works for models officially supported, it will not work when falling back on `transformers`. You can get more information about unsupported models [here](./basic_tutorials/non_core_models.md).

Reviewer 3: ```suggestion
Tensor Parallelism only works for [models officially supported](../supported_models), it will not work when falling back to `transformers`. You can get more information about unsupported models [here](../basic_tutorials/non_core_models).
```

That page does not exist yet, does it?

Author: Yep but Omar wanted me to add it so I added that

Author: Also will change the title.

Author: it exists here no?: https://huggingface.co/docs/text-generation-inference/supported_models

Author: Yes, Omar wanted me to add it that's why I added it nevertheless.

</Tip>

> You can learn more in-depth about tensor-parallelism from `transformers` docs in this [link](https://huggingface.co/docs/transformers/main/en/perf_train_gpu_many#tensor-parallelism).

Reviewer 3: ```suggestion
You can learn a lot more details about tensor-parallelism from [the `transformers` docs](https://huggingface.co/docs/transformers/main/en/perf_train_gpu_many#tensor-parallelism).
```

# Tensor Parallelism

> Tensor parallelism (also called horizontal model parallelism) is a technique used to fit a large model in multiple GPUs.  Intermediate outputs between ranks are sent and received from one rank to another in a synchronous or asynchronous manner. When multiplying input with weights for inference, multiplying input with weights directly is equivalent to dividing the weight matrix column-wise, multiplying each column with input separately, and then concatenating the separate outputs like below 👇 

Reviewer: ```suggestion
Tensor parallelism is a technique used to fit a large model in multiple GPUs.  Intermediate outputs between GPUs are sent and received from one GPU to another in a synchronous or asynchronous manner. For example, when multiplying the input tensors with the first weights tensor, multiplying both tensors is equivalent to splitting the weight tensor column-wise, multiplying each column with input separately, and then concatenating the separate outputs like below 👇 
```

> Tensor parallelism (also called horizontal model parallelism) is a technique used to fit a large model in multiple GPUs.  Intermediate outputs between ranks are sent and received from one rank to another in a synchronous or asynchronous manner. When multiplying input with weights for inference, multiplying input with weights directly is equivalent to dividing the weight matrix column-wise, multiplying each column with input separately, and then concatenating the separate outputs like below 👇 

Reviewer: I think that the wording could improve a bit
* is the synchronous or asynchronous part important for explaining TP?
* you say the intermediate outputs. I'm not sure this is very clear
* TP distributes chunks of layers across multiple GPUs. This is not mentioned/clear in the definition

Author: I addressed rest of your comments.

> ![Image courtesy of Anton Lozkhov](https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/tgi/TP.png)

Reviewer: I think the image is slightly misaligned. The bottom part seems more to the right

Author: I edited it to look better, but github preview will not reflect that.

Reviewer 3: Omar is right, the bottom box is ~10 pixels more to the right (just downloaded from `documentation-images`)

Author: I fixed it.

> In TGI, tensor parallelism is implemented under the hood by sharding weights and placing them in different GPUs. The matrix multiplications then take place in different GPUs and are then gathered into a single tensor. 

Reviewer: Maybe this is a bit redundant, as this is repeating what was said in the initial paragraph

Author: I addressed this.

> In TGI, tensor parallelism is implemented under the hood by sharding weights and placing them in different ranks. The matrix multiplications then take place in different ranks and are then gathered into a single tensor. 

Reviewer: ```suggestion
In TGI, tensor parallelism is implemented under the hood by sharding weights and placing them in different GPUs. The matrix multiplications then take place in different GPUs and are then gathered into a single tensor. 
```

Author: I thought it would be useful to give context since when you serve, in logs, it says e.g. `shard is ready at rank=0`

<Tip warning={true}>

> Tensor Parallelism only works for models officially supported, it will not work when falling back on `transformers`.

Reviewer: Maybe link to the part that explains custom models once that's merged

Reviewer: This is unresolved I think

Author: It's not merged yet

Author: I addressed this nevertheless.

</Tip>

> You can learn more in-depth on tensor-parallelism from transformers docs in this [link](https://huggingface.co/docs/transformers/main/en/perf_train_gpu_many#tensor-parallelism).

Reviewer: ```suggestion
You can learn more in-depth about tensor-parallelism from `transformers` docs in this [link](https://huggingface.co/docs/transformers/main/en/perf_train_gpu_many#tensor-parallelism).
```
